make/photon/db/docker-entrypoint.sh

```diff
@@ -89,26 +101,61 @@ EOSQL
         psql+=( --username "$POSTGRES_USER" --dbname "$POSTGRES_DB" )
 
         echo
-        for f in /docker-entrypoint-initdb.d/*; do
-                case "$f" in
-                        *.sh)     echo "$0: running $f"; . "$f" ;;
-                        *.sql)    echo "$0: running $f"; "${psql[@]}" -f "$f"; echo ;;
-                        *.sql.gz) echo "$0: running $f"; gunzip -c "$f" | "${psql[@]}"; echo ;;
-                        *)        echo "$0: ignoring $f" ;;
-                esac
-                echo
-        done
+
+        if [ $2 == "true" ]; then
+                for f in /docker-entrypoint-initdb.d/*; do
+                        case "$f" in
+                                *.sh)     echo "$0: running $f"; . "$f" ;;
+                                *.sql)    echo "$0: running $f"; "${psql[@]}" -f "$f"; echo ;;
+                                *.sql.gz) echo "$0: running $f"; gunzip -c "$f" | "${psql[@]}"; echo ;;
+                                *)        echo "$0: ignoring $f" ;;
+                        esac
+                        echo
+                done
+        fi
 
         PGUSER="${PGUSER:-postgres}" \
-        pg_ctl -D "$PGDATA" -m fast -w stop
+        pg_ctl -D "$1" -m fast -w stop
 
         echo
         echo 'PostgreSQL init process complete; ready for start up.'
         echo
+
+}
+
+# look specifically for PG_VERSION, if exist and needs upgrade, use pg_upgrade to handle it. Otherwise just init db.
+if [ -s $PGDATA/PG_VERSION ]; then
+        PG_VERSION_OLD=$(cat $PGDATA/PG_VERSION);
+        if [ "`echo "${PG_VERSION_OLD} < $PG_VERSION_NEW" | bc`" -eq 1 ]; then
+                echo 'PostgreSQL start to move $PGDATA to $PGDATAOLD.'
+                mkdir -p $PGDATAOLD
+                chown -R postgres "$PGDATAOLD"
+                mv -f $PGDATA/* $PGDATAOLD
+                chmod 700 $PGDATAOLD
+
+                echo 'init new DB.'
+                initPG $PGDATANEW false
+
+                echo 'execute the pg_upgrade.'
+                cd /tmp && /usr/bin/pg_upgrade \
+                  --old-datadir=$PGDATAOLD \
+                  --new-datadir=$PGDATANEW \
+                  --old-bindir=$PGBINOLD \
+                  --new-bindir=$PGBINNEW \
+                  --old-options '-c config_file=$PGDATAOLD/postgresql.conf' \
+                  --new-options '-c config_file=$PGDATANEW/postgresql.conf'
+                mv $PGDATAOLD/pg_hba.conf $PGDATANEW/pg_hba.conf
+
+                echo 'PostgreSQL start to move $PGDATANEW to $PGDATA.'
+                mv -f $PGDATANEW/* $PGDATA
+                rm -rf $PGDATAOLD;
+                rm -rf $PGDATANEW;
+        fi
+else
+        initPG $PGDATA true
```

Comment:
Could add some comments for arguments of the initPG function? Especially the second argument.

Reply:
added